uPortSDK/Classes/EthrDID/EthrDIDResolver.swift

```diff
@@ -0,0 +1,247 @@
+                    default:
+                        print( "unknown delegateType -> \(delegateType)")
+                    }
+                }
+            }
+            
+            if baseEvent is EthereumDIDRegistry.Events.DIDAttributeChanged.Arguments {
+                let event = baseEvent as! EthereumDIDRegistry.Events.DIDAttributeChanged.Arguments
+                let validTo = Int64( event.validto.value )
+                if now <= validTo {
+                    let nameOptional = String( data: event.name.value, encoding: .utf8 )
+                    
+                    guard let name = nameOptional else {
+                        print( "invalid DIDAttributeChanged event name" )
+                        continue
+                    }
+                    
+                    let key = "DIDAttributeChanged-\(name)-\(event.value.value.hexEncodedString())"
+                    let regex = try? NSRegularExpression( pattern: "^did/(pub|auth|svc)/(\\w+)(/(\\w+))?(/(\\w+))?$", options: .caseInsensitive )
+                    let matchResult = regex?.matches(in: name, options: [], range: NSRange(location: 0, length: name.count))
+                    guard let textCheckingResult = matchResult?.first else {
+                        continue
+                    }
+                    
+                    var matches
+
+                }
+
+            }
+            
+            
+        }
+        
+        return DDO(id: "TODO: implement")
+    }
+    
+    func lastChanged( identity: String ) -> Promise<String> {
+        return Promise<String> { fulfill, reject in
+            guard let address = BigUInt( identity.withoutHexPrefix, radix: 16 ),
+                    let solidityAddress = try? Solidity.Address( bigUInt: address ) else {
+                reject( EthrDIDResolverError.invalidIdentity )
+                return
+            }
+            
+            let encodedCall = EthereumDIDRegistry.Changed.encodeCall( arguments: solidityAddress )
+            self.rpc.ethCall(address: self.registryAddress, data: encodedCall).then({ response in
+                guard let responseUnwrapped = response else {
+                    reject( EthrDIDResolverError.invalidRPCResponse )
+                    return
+                }
+                
+                let parsedResponse = JsonRpcBaseResponse.fromJson(json: responseUnwrapped)
+                guard parsedResponse.error == nil || parsedResponse.result != nil else {
+                    reject( EthrDIDResolverError.invalidRPCResponse )
+                    return
+                }
+                
+                fulfill( "\(parsedResponse.result!)")
+            }).catch({ error in
+                reject( error )
+            })
+        }
+    }
+    
+    func lastChangedSynchronous( identity: String ) throws -> String {
```

Comment:
Dumb question. These functions marked Synchronous. Are they actually synchronous? Do they stall the app when run?

Reply:
Correct! Synchronous here means they will stall the thread/queue they are called on until the network call completes. Synchronous methods are meant to be called on a background thread/queue.